src/items/powerup_manager.cpp

```diff
@@ -205,142 +208,343 @@ void PowerupManager::LoadPowerup(PowerupType type, const XMLNode &node)
 }   // LoadNode
 
 // ----------------------------------------------------------------------------
-/** Loads a weight list specified in powerup.xml. The different position
- *  classes must be loaded in the right order
+/** Loads a weight list specified in powerup.xml.
+ *  Calculates a linear average of the weight specified for the reference points
+ *  of num_karts higher and lower (unless it exactly matches one)
  *  \param root The root node of powerup.xml
+ *  \param num_karts The number of karts in the race
  *  \param class_name The name of the position class to load.
  *  \param position_class The class for which the weights are read.
  */
 void PowerupManager::loadWeights(const XMLNode &root,
+                                 unsigned int num_karts,
                                  const std::string &class_name,
                                  PositionClass position_class)
 {
-    const XMLNode *node = root.getNode(class_name);
-    std::string s(""), s_multi("");
-    if(node) node->get("w",       &s      );
-    if(node) node->get("w-multi", &s_multi);
+    const XMLNode *node = root.getNode(class_name), *node2 = root.getNode("reference_points");
+    std::string s_inf(""), s_inf_multi(""), s_sup(""), s_sup_multi(""), ref;
+    std::string w_inf("w"), w_sup("w"), w_inf_multi("w-multi"), w_sup_multi("w-multi");
+    int inf_num = 1, sup_num = 1; //The number of karts associated with the reference points
+
+    /** Adds to w_inf and w_sup the suffixe of the closest num_karts reference classes
+     */
+    if(position_class!=POSITION_BATTLE_MODE 
+    && position_class!=POSITION_TUTORIAL_MODE
+    && position_class!=POSITION_SOCCER_MODE)
+    {
+        //First step : get the reference points
+
+        if(!node2)
+        {
+            Log::error("[PowerupManager]","No reference points found, "
+                         "probabilities will be incorrect");
+        }
+
+        if (node2)
+        {
+            if(race_manager->isFollowMode())
+            {
+                w_inf = w_sup = "wf";
+            }
+            
+            else
+            {
+                // With default config it won't loop more than five time
+                // This is mainly to let full config editing power
+                for(int i=1;i<=50;i++)
+                {
+                    ref = "ref" + StringUtils::toString(i);
+                    std::string ref_pos;
+                    if (!node2->get(ref, &ref_pos))
+                        break;
+                    unsigned int ref_value = atoi(ref_pos.c_str());
```

Comment:
atoi should be avoided as it has very peculiar error handling. Take a look at StingUtils::parseString<int> (will require a bit more error handling but that's a good thing)

Reply:
The original code also used atoi. But it seems to work with parseString and there are now error warnings if it goes wrong, so it's better I suppose.